src/deserialize.rs

```diff
@@ -173,8 +175,52 @@ impl_arrow_array!(Utf8Array<i32>);
 impl_arrow_array!(BinaryArray<i32>);
 impl_arrow_array!(ListArray<i32>);
 
-/// Helper method to return an iterator over [`T`] from a boxed arrow2 Array
-pub fn arrow_array_typed_iterator<'a, T>(b: &'a dyn Array) -> arrow2::error::Result<impl Iterator<Item = T> + 'a>
+
+/// Top-level API to deserialize from Arrow
+pub trait FromArrow<T>
+{
+    fn from_arrow(self) -> arrow2::error::Result<T>;
+}
+
+impl<'a,T> FromArrow<Vec<T>> for &'a dyn Array 
+where T: ArrowDeserialize + 'static, 
+    for<'b> &'b <T as ArrowDeserialize>::ArrayType: IntoIterator,
+{
+    fn from_arrow(self) -> arrow2::error::Result<Vec<T>> {
+        if &<T as ArrowField>::data_type() != self.data_type() {
+            Err(arrow2::error::ArrowError::InvalidArgumentError("Data type mismatch".to_string()))
+        }
+        else {
+            Ok(arrow_array_deserialize_unchecked_iter(self)?.collect())
+        }
+    }
+}
+
+// Blanket implemented for Box<dyn Array>
+impl<'a, T> FromArrow<T> for &'a Box<dyn Array>
+where &'a dyn Array: FromArrow<T>
+{
+    fn from_arrow(self) -> arrow2::error::Result<T> {
+        use std::ops::Deref;
+
+        self.deref().from_arrow()
+    }
+}
+
+// Blanket implemented for Arc<dyn Array>
+impl<'a, T> FromArrow<T> for &'a std::sync::Arc<dyn Array>
+where &'a dyn Array: FromArrow<T>
+{
+    fn from_arrow(self) -> arrow2::error::Result<T> {
+        use std::ops::Deref;
+
+        self.deref().from_arrow()
+    }
+}
+
+/// Returns an iterator for elements from a [`arrow2::array::Array`]
+/// Does not perform any schema checks.
+pub fn arrow_array_deserialize_unchecked_iter<'a, T>(b: &'a dyn arrow2::array::Array) -> arrow2::error::Result<impl Iterator<Item = T> + 'a>
```

Comment:
`unchecked` is usually reserved for functions that are `unsafe`. Maybe `schemaless` or `untyped`?

Reply:
I have another commit on top of this that addressed nested optional structs. I'll address this there.